File: dna_score_mutations.R
# Required libraries
library(Biostrings)
library(pwalign)
library(parallel)

args <- commandArgs(trailingOnly = TRUE)
input_file <- args[1]         # e.g. "<sample>.prot.input.tsv"
threads    <- as.integer(args[2])

calculate_mutation_score <- function(reference, target, changes_str, debug = FALSE) {
  if (is.null(changes_str) || changes_str == "" || is.na(changes_str)) {
    return(list(score = 0, detected_mutations = "None",
                wt_confirmed_positions = 0, status = "Unknown"))
  }

  # Parse tokens like: C12A, T22G, G39GG, T40TGT, C43T, G43-, etc.
  change_tokens <- trimws(strsplit(changes_str, ",")[[1]])
  parse_token <- function(tok) {
    tok <- toupper(tok)
    m <- regexec("^([ACGT]+)([0-9]+)([ACGT\\-]+)$", tok)
    mm <- regmatches(tok, m)[[1]]
    if (length(mm) != 4) return(NULL)
    list(RefSeg = mm[2], Pos = as.integer(mm[3]), AltSeg = mm[4], Raw = tok)
  }
  parsed <- lapply(change_tokens, parse_token)
  parsed <- parsed[!vapply(parsed, is.null, logical(1))]
  if (length(parsed) == 0) {
    return(list(score = 0, detected_mutations = "None",
                wt_confirmed_positions = 0, status = "Unknown"))
  }

  # Normalize inputs
  reference <- toupper(gsub("[ \r\n\t]", "", reference))
  target    <- toupper(gsub("[- \r\n\t]", "", target))

  # Align: force full reference (pattern) to align; read (subject) can be local
  submat <- nucleotideSubstitutionMatrix(match = 2, mismatch = -5, baseOnly = FALSE)  # tolerate N/IUPAC
  pa <- pairwiseAlignment(
    pattern = DNAString(reference),
    subject = DNAString(target),
    type = "global-local",
    substitutionMatrix = submat,
    gapOpening = -10, gapExtension = -0.5
  )

  aligned_ref    <- as.character(alignedPattern(pa))
  aligned_target <- as.character(alignedSubject(pa))
  ref_chars    <- strsplit(aligned_ref, "")[[1]]
  target_chars <- strsplit(aligned_target, "")[[1]]

  # Build map: alignment column -> (ungapped) reference position
  ref_pos_at_col <- rep(NA_integer_, length(ref_chars))
  rp <- 0L
  for (k in seq_along(ref_chars)) {
    if (ref_chars[k] != "-") {
      rp <- rp + 1L
      ref_pos_at_col[k] <- rp
    }
  }
  max_ref_pos <- max(ref_pos_at_col, na.rm = TRUE)

  informative <- function(b) !(b %in% c("-", "N"))

  # Subject coverage window (columns where subject is not a gap)
  covered_cols <- which(target_chars != "-")
  if (length(covered_cols) == 0L) {
    return(list(score = 0, detected_mutations = "None",
                wt_confirmed_positions = 0, status = "Unknown"))
  }
  cov_start <- min(covered_cols); cov_end <- max(covered_cols)
  inside_cov <- function(cols_or_j) {
    if (length(cols_or_j) == 1L) {
      j <- cols_or_j
      !is.na(j) && j >= cov_start && j <= cov_end
    } else {
      length(cols_or_j) > 0L && min(cols_or_j) >= cov_start && max(cols_or_j) <= cov_end
    }
  }

  take_at_cols <- function(chars, cols) paste(chars[cols], collapse = "")

  score <- 0L
  detected <- character(0)

  # Track unique positions for WT confirmation and detected calls
  detected_pos <- integer(0)
  wt_pos <- integer(0)
  add_wt_once <- function(pos) {
    if (length(pos) == 1L && !(pos %in% detected_pos) && !(pos %in% wt_pos)) {
      wt_pos <<- c(wt_pos, pos)
    }
  }

  for (p in parsed) {
    pos     <- p$Pos
    ref_seg <- p$RefSeg
    alt_seg <- p$AltSeg
    raw_tok <- p$Raw

    ref_len <- nchar(ref_seg)
    alt_len <- nchar(alt_seg)

    # Sanity on reference coordinate range
    if (is.na(pos) || pos < 1 || (pos + ref_len - 1) > max_ref_pos) next

    # Alignment columns spanned by the reference segment
    ref_cols <- which(ref_pos_at_col %in% pos:(pos + ref_len - 1))
    if (length(ref_cols) != ref_len) next  # ambiguous/misaligned
    if (!inside_cov(ref_cols)) next        # coverage guard

    ref_block <- take_at_cols(ref_chars, ref_cols)
    read_block <- take_at_cols(target_chars, ref_cols)

    # Reference spec must match alignment's reference block
    if (ref_block != ref_seg) next

    # ---- CASE A: Deletion ----
    if (alt_seg == "-") {
      gaps <- vapply(ref_cols, function(j) target_chars[j] == "-", logical(1))
      if (all(gaps)) {
        score <- score + 1L
        detected <- c(detected, paste0(ref_seg, pos, "-"))
        detected_pos <- c(detected_pos, pos)
      } else if (read_block == ref_seg && !grepl("N", read_block)) {
        add_wt_once(pos)
      }
      next
    }

    # ---- CASE B: Insertion-after (e.g., G39GG, T40TGT) ----
    if (ref_len == 1 && alt_len > 1 && startsWith(alt_seg, ref_seg)) {
      j <- ref_cols[1]
      base_ok <- (target_chars[j] == ref_seg)
      alt_tail <- substring(alt_seg, 2)            # bases inserted after the ref base
      tail_needed <- nchar(alt_tail)

      # Walk right collecting subject bases where reference has gaps
      jj <- j + 1
      tail_seen <- character(0)
      started <- FALSE
      while (jj <= length(ref_chars) && nchar(paste0(tail_seen, collapse = "")) < tail_needed) {
        if (ref_chars[jj] == "-") {
          aa <- target_chars[jj]
          if (informative(aa)) {
            tail_seen <- c(tail_seen, aa)
            started <- TRUE
          }
        } else {
          if (started) break  # stop at first ref letter after starting insertion tract
        }
        jj <- jj + 1
      }
      tail_seq <- paste0(tail_seen, collapse = "")
      if (base_ok && tail_seq == alt_tail) {
        score <- score + 1L
        detected <- c(detected, raw_tok)
        detected_pos <- c(detected_pos, pos)
      } else if (base_ok) {
        add_wt_once(pos)
      }
      next
    }

    # ---- CASE C: Equal-length replacement (incl. multi-base) ----
    if (ref_len == alt_len && alt_seg != ref_seg) {
      if (!grepl("[-N]", read_block) && read_block == alt_seg) {
        score <- score + 1L
        detected <- c(detected, raw_tok)
        detected_pos <- c(detected_pos, pos)
      } else if (!grepl("[-N]", read_block) && read_block == ref_seg) {
        add_wt_once(pos)
      }
      next
    }

    # ---- CASE D: Single-base substitution fallback ----
    if (ref_len == 1 && alt_len == 1 && alt_seg != ref_seg) {
      j <- ref_cols[1]
      rj <- ref_chars[j]
      sj <- target_chars[j]
      if (informative(sj)) {
        if (rj == ref_seg && sj == alt_seg) {
          score <- score + 1L
          detected <- c(detected, paste0(ref_seg, pos, alt_seg))
          detected_pos <- c(detected_pos, pos)
        } else if (rj == ref_seg && sj == ref_seg) {
          add_wt_once(pos)
        }
      }
      next
    }

    # ---- CASE E: Nothing matched → possible WT confirmation ----
    if (!grepl("[-N]", read_block) && read_block == ref_seg) {
      add_wt_once(pos)
    }
  }

  detected_str <- if (length(detected) > 0) paste(detected, collapse = ",") else "None"
  wt_confirmed <- length(unique(wt_pos))
  status <- if (score > 0) "Resistant" else if (wt_confirmed > 0) "Wildtype" else "Unknown"

  list(score = score,
       detected_mutations = detected_str,
       wt_confirmed_positions = wt_confirmed,
       status = status)
}

process_mutation_data <- function(file_path) {
  input_data <- read.delim(file_path,
                           sep = "\t",
                           stringsAsFactors = FALSE,
                           quote = "",
                           check.names = FALSE)

  required_cols <- c("class", "gene", "read", "reference", "target", "changes_str")
  missing_cols <- required_cols[!required_cols %in% colnames(input_data)]
  if (length(missing_cols) > 0) {
    stop(paste("Missing required columns:", paste(missing_cols, collapse = ", ")))
  }

#  results <- lapply(seq_len(nrow(input_data)), function(i) {
  results <- mclapply(seq_len(nrow(input_data)), function(i) {
    calculate_mutation_score(
      reference   = input_data$reference[i],
      target      = input_data$target[i],
      changes_str = input_data$changes_str[i]
    )
  }, mc.cores = threads)

  input_data$MutationScore         <- sapply(results, `[[`, "score")
  input_data$DetectedMutations     <- sapply(results, `[[`, "detected_mutations")
  input_data$WTConfirmedPositions  <- sapply(results, `[[`, "wt_confirmed_positions")
  input_data$Status                <- sapply(results, `[[`, "status")

  write.table(input_data,
              "updated_table_with_scores_and_mutations.tsv",
              sep = "\t",
              row.names = FALSE,
              quote = FALSE)

  input_data
}

# Run with the same workflow as before
result <- process_mutation_data(input_file)
warnings()
print(result)

File: dna_score_mutations.py
#!/usr/bin/env python3
"""Score DNA mutations from a tab-separated MetaPointFinder input file.

Python translation of ``dna_score_mutations.R``.  The reference is aligned
globally to a local region of the target, using the same match, mismatch and
affine-gap scores as the R implementation.
"""

from __future__ import annotations

import argparse
import csv
import math
import re
from concurrent.futures import ProcessPoolExecutor
from dataclasses import dataclass
from pathlib import Path


OUTPUT_FILE = "updated_table_with_scores_and_mutations.tsv"
TOKEN_RE = re.compile(r"^([ACGT]+)([0-9]+)([ACGT-]+)$")
NEG_INF = -math.inf


@dataclass(frozen=True)
class Mutation:
    ref: str
    position: int
    alt: str
    raw: str


def _global_local_align(reference: str, target: str) -> tuple[str, str]:
    """Globally align reference to a local target region with affine gaps."""
    m, n = len(reference), len(target)
    # M: paired bases; X: reference base/target gap; Y: reference gap/target base.
    matrices = [[[NEG_INF] * (n + 1) for _ in range(m + 1)] for _ in range(3)]
    traces = [[[-1] * (n + 1) for _ in range(m + 1)] for _ in range(3)]
    match, mismatch, gap_open, gap_extend = 2.0, -5.0, -10.0, -0.5

    matrices[0][0][0] = 0.0
    # A target prefix is free (local with respect to the subject).
    for j in range(1, n + 1):
        matrices[0][0][j] = 0.0
        traces[0][0][j] = 0
    for i in range(1, m + 1):
        matrices[1][i][0] = gap_open + (i - 1) * gap_extend
        traces[1][i][0] = 0 if i == 1 else 1

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            candidates = [matrices[s][i - 1][j - 1] for s in range(3)]
            previous = max(range(3), key=candidates.__getitem__)
            score = match if reference[i - 1] == target[j - 1] else mismatch
            matrices[0][i][j] = candidates[previous] + score
            traces[0][i][j] = previous

            candidates = [
                matrices[0][i - 1][j] + gap_open,
                matrices[1][i - 1][j] + gap_extend,
                matrices[2][i - 1][j] + gap_open,
            ]
            previous = max(range(3), key=candidates.__getitem__)
            matrices[1][i][j] = candidates[previous]
            traces[1][i][j] = previous

            candidates = [
                matrices[0][i][j - 1] + gap_open,
                matrices[1][i][j - 1] + gap_open,
                matrices[2][i][j - 1] + gap_extend,
            ]
            previous = max(range(3), key=candidates.__getitem__)
            matrices[2][i][j] = candidates[previous]
            traces[2][i][j] = previous

    state, end_j = max(
        ((s, j) for s in range(3) for j in range(n + 1)),
        key=lambda item: matrices[item[0]][m][item[1]],
    )
    i, j = m, end_j
    aligned_ref: list[str] = []
    aligned_target: list[str] = []
    while i > 0:
        previous = traces[state][i][j]
        if state == 0:
            aligned_ref.append(reference[i - 1])
            aligned_target.append(target[j - 1])
            i -= 1
            j -= 1
        elif state == 1:
            aligned_ref.append(reference[i - 1])
            aligned_target.append("-")
            i -= 1
        else:
            aligned_ref.append("-")
            aligned_target.append(target[j - 1])
            j -= 1
        state = previous
    return "".join(reversed(aligned_ref)), "".join(reversed(aligned_target))


def _unknown() -> dict[str, object]:
    return {
        "score": 0,
        "detected_mutations": "None",
        "wt_confirmed_positions": 0,
        "status": "Unknown",
    }


def calculate_mutation_score(
    reference: str | None, target: str | None, changes_str: str | None
) -> dict[str, object]:
    if not changes_str or changes_str.strip().upper() in {"", "NA", "NAN"}:
        return _unknown()

    mutations: list[Mutation] = []
    for token in changes_str.split(","):
        raw = token.strip().upper()
        match = TOKEN_RE.fullmatch(raw)
        if match:
            mutations.append(Mutation(match[1], int(match[2]), match[3], raw))
    if not mutations:
        return _unknown()

    reference = re.sub(r"\s", "", reference or "").upper()
    target = re.sub(r"[-\s]", "", target or "").upper()
    if not reference or not target:
        return _unknown()
    aligned_ref, aligned_target = _global_local_align(reference, target)

    ref_positions: list[int | None] = []
    ref_position = 0
    for base in aligned_ref:
        if base != "-":
            ref_position += 1
            ref_positions.append(ref_position)
        else:
            ref_positions.append(None)

    covered = [i for i, base in enumerate(aligned_target) if base != "-"]
    if not covered:
        return _unknown()
    coverage_start, coverage_end = min(covered), max(covered)
    score = 0
    detected: list[str] = []
    detected_positions: set[int] = set()
    wt_positions: set[int] = set()

    def add_wt(position: int) -> None:
        if position not in detected_positions:
            wt_positions.add(position)

    for mutation in mutations:
        pos, ref_seg, alt_seg = mutation.position, mutation.ref, mutation.alt
        ref_len, alt_len = len(ref_seg), len(alt_seg)
        if pos < 1 or pos + ref_len - 1 > ref_position:
            continue
        wanted = set(range(pos, pos + ref_len))
        ref_cols = [i for i, value in enumerate(ref_positions) if value in wanted]
        if (len(ref_cols) != ref_len or min(ref_cols) < coverage_start
                or max(ref_cols) > coverage_end):
            continue
        ref_block = "".join(aligned_ref[i] for i in ref_cols)
        read_block = "".join(aligned_target[i] for i in ref_cols)
        if ref_block != ref_seg:
            continue

        found = False
        if alt_seg == "-":
            found = all(aligned_target[i] == "-" for i in ref_cols)
            if not found and read_block == ref_seg and "N" not in read_block:
                add_wt(pos)
        elif ref_len == 1 and alt_len > 1 and alt_seg.startswith(ref_seg):
            col = ref_cols[0]
            base_ok = aligned_target[col] == ref_seg
            needed = alt_seg[1:]
            seen: list[str] = []
            started = False
            next_col = col + 1
            while next_col < len(aligned_ref) and len(seen) < len(needed):
                if aligned_ref[next_col] == "-":
                    base = aligned_target[next_col]
                    if base not in {"-", "N"}:
                        seen.append(base)
                        started = True
                elif started:
                    break
                next_col += 1
            found = base_ok and "".join(seen) == needed
            if not found and base_ok:
                add_wt(pos)
        elif ref_len == alt_len and alt_seg != ref_seg:
            found = "-" not in read_block and "N" not in read_block and read_block == alt_seg
            if not found and "-" not in read_block and "N" not in read_block and read_block == ref_seg:
                add_wt(pos)
        elif "-" not in read_block and "N" not in read_block and read_block == ref_seg:
            add_wt(pos)

        if found:
            score += 1
            detected.append(ref_seg + str(pos) + "-" if alt_seg == "-" else mutation.raw)
            detected_positions.add(pos)
            wt_positions.discard(pos)

    return {
        "score": score,
        "detected_mutations": ",".join(detected) if detected else "None",
        "wt_confirmed_positions": len(wt_positions),
        "status": "Resistant" if score else ("Wildtype" if wt_positions else "Unknown"),
    }


def _score_row(row: dict[str, str]) -> dict[str, object]:
    return calculate_mutation_score(row["reference"], row["target"], row["changes_str"])


def process_mutation_data(file_path: str | Path, threads: int = 1) -> list[dict[str, str]]:
    with Path(file_path).open(newline="", encoding="utf-8-sig") as handle:
        reader = csv.DictReader(handle, delimiter="\t", quoting=csv.QUOTE_NONE)
        required = {"class", "gene", "read", "reference", "target", "changes_str"}
        missing = required.difference(reader.fieldnames or [])
        if missing:
            raise ValueError(f"Missing required columns: {', '.join(sorted(missing))}")
        rows = list(reader)
        fieldnames = list(reader.fieldnames or [])

    if threads > 1 and len(rows) > 1:
        with ProcessPoolExecutor(max_workers=threads) as executor:
            results = list(executor.map(_score_row, rows))
    else:
        results = [_score_row(row) for row in rows]

    output_columns = [
        "MutationScore", "DetectedMutations", "WTConfirmedPositions", "Status"
    ]
    for row, result in zip(rows, results):
        row.update(dict(zip(output_columns, result.values())))
    with Path(OUTPUT_FILE).open("w", newline="", encoding="utf-8") as handle:
        writer = csv.DictWriter(
            handle, fieldnames=fieldnames + output_columns, delimiter="\t",
            quoting=csv.QUOTE_NONE, lineterminator="\n",
        )
        writer.writeheader()
        writer.writerows(rows)
    return rows


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("input_file", help="Input TSV file")
    parser.add_argument("threads", nargs="?", type=int, default=1)
    args = parser.parse_args()
    rows = process_mutation_data(args.input_file, max(1, args.threads))
    for row in rows:
        print(row)


if __name__ == "__main__":
    main()
